core/src/main/scala/kafka/server/AdminManager.scala

```diff
@@ -255,14 +258,28 @@ class AdminManager(val config: KafkaConfig,
         resource.`type` match {
           case ResourceType.TOPIC =>
             val topic = resource.name
+
             val properties = new Properties
             config.entries.asScala.foreach { configEntry =>
               properties.setProperty(configEntry.name(), configEntry.value())
             }
-            if (validateOnly)
-              AdminUtils.validateTopicConfig(zkUtils, topic, properties)
-            else
-              AdminUtils.changeTopicConfig(zkUtils, topic, properties)
+
+            alterConfigPolicy match {
+              case Some(policy) =>
+                AdminUtils.validateTopicConfig(zkUtils, topic, properties)
+
+                val configEntriesMap = config.entries.asScala.map(entry => (entry.name, entry.value)).toMap
+                policy.validate(new AlterConfigPolicy.RequestMetadata(
+                  new ConfigResource(ConfigResource.Type.TOPIC, resource.name), configEntriesMap.asJava))
+
+                if (!validateOnly)
+                  AdminUtils.changeTopicConfig(zkUtils, topic, properties)
```

Comment:
This does `validateTopicConfig` again? The code looks odd where validate is done when policy is set, but not otherwise. But actually it does validate twice when policy is set. Maybe it is ok if validation is very low overhead.